ifrs17/Utils/ActivityLogsAndImports.ipynb

```diff
@@ -0,0 +1,302 @@
+            "source": [
+                "public static async Task<ActivityLog> ExecuteWithStoreActivityAsync(this ImportOptionsBuilder builder, ",
+                "\n                                                                                ISessionVariable session, ",
+                "\n                                                                                IDataSource dataSource)",
+                "\n{",
+                "\n    var log = await builder.ExecuteAsync();",
+                "\n    var options = builder.GetImportOptions();",
+                "\n    var activity = new KeyedActivity(log, session);",
+                "\n    switch(options)",
+                "\n    {",
+                "\n        case FileImportOptions fio:",
+                "\n            var keyedFile = await (new KeyedFile(fio)).InitializeImportDataAsync(session);",
+                "\n            activity = activity with {SourceId = keyedFile.Id, ",
+                "\n                                        Category = \"Import from File\"};",
+                "\n            if (keyedFile is KeyedFile kf) await dataSource.UpdateAsync<KeyedFile>(kf.RepeatOnce());",
+                "\n            break;",
+                "\n        case StringImportOptions sgio:",
+                "\n            var keyedStringImport = await (new KeyedStringImport(sgio)).InitializeImportDataAsync(session);",
+                "\n            activity = activity with {SourceId = keyedStringImport.Id, ",
+                "\n                                    Category = \"Import from String\"};",
+                "\n            if (keyedStringImport is KeyedStringImport ksi) await dataSource.UpdateAsync<KeyedStringImport>(ksi.RepeatOnce());",
+                "\n            break;",
+                "\n        case StreamImportOptions smio:",
+                "\n            var keyedStreamImport = await (new KeyedStreamImport(smio)).InitializeImportDataAsync(session);",
+                "\n            activity = activity with {SourceId = keyedStreamImport.Id, ",
+                "\n                                    Category = \"Import from Stream\"};",
+                "\n            if (keyedStreamImport is KeyedStreamImport ksti) await dataSource.UpdateAsync<KeyedStreamImport>(ksti.RepeatOnce());",
+                "\n            break;",
+                "\n        case DataSetImportOptions dsio:",
+                "\n            var keyedDataSetImport = await (new KeyedDataSetImport(dsio)).InitializeImportDataAsync(session);",
+                "\n            activity = activity with {SourceId = keyedDataSetImport.Id, ",
+                "\n                                    Category = \"Import from Data Set\"};",
+                "\n            if (keyedDataSetImport is KeyedDataSetImport kdsi) await dataSource.UpdateAsync<KeyedDataSetImport>(kdsi.RepeatOnce());",
+                "\n            break;",
+                "\n        default:",
```

Comment:
Is this default (doing nothing in case nothing matches) correct? I don't know, but is this even possible (to end up here)?
If not, maybe create an Error.

Reply:
Yes, it throws an exception now.